### PR TITLE
Improve baseline correction with feature resolver

### PR DESCRIPTION
**Pull request by robinschmid**
contains #2126

- Reuse some logic

Feature resolver struggles with high baseline and uses chromatographic threshold to remove data points. This previously led to the yBuffer to be changed. So make a defensive copy and apply baseline correction first before feature detection.

Steps are now:

1. baseline correction on copy of data
2. feature resolver on this copy
3. exclude ranges from orginal data
4. ranges define landmark indices of interest that should be used when subsampling data points from the filtered data. This ensures that there are samples right next to features.
5. baseline correction on original data

image

### CODE REVIEW

File: mzmine-community/src/test/java/util/SplineBaselineCorrectorTest.java [hunk 1]
```diff
@@ -25,83 +25,110 @@
...
+    float[] floats = ArrayUtils.doubleToFloat(src);
+    var data = new SimpleOtherTimeSeries(null, floats, src, "test", otherData);
+    BaselineDataBuffer buffer = new BaselineDataBuffer();
+    buffer.extractDataIntoBuffer(data);
+    return buffer;
+  }
+
+  private void check(final BaselineDataBuffer buffer, final int expectedRemaining,
+      final double[] expectedX, final int[] expectedIndicesOfInterest) {
+    logger.info(Arrays.stream(buffer.xBufferRemovedPeaks()).mapToObj(String::valueOf)
```
Comment (SteffenHeu) [inline mzmine-community/src/test/java/util/SplineBaselineCorrectorTest.java:70]: debugging leftover?

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/UnivariateBaselineCorrector.java [hunk 2]
```diff
@@ -50,57 +50,75 @@ public UnivariateBaselineCorrector(@Nullable MemoryMapStorage storage, int numSa
...
-    return createNewTimeSeries(timeSeries, numValues, yBuffer);
+      additionalData.add(
+          new AnyXYProvider(Color.BLUE, "samples", subData.numValues(), subData::getX,
+              subData::getY));
+    }
   }
 
+
   @Override
   public <T extends IntensityTimeSeries> T correctBaseline(T timeSeries) {
```
Comment (SteffenHeu) [inline mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/UnivariateBaselineCorrector.java:94]: is this still different from the AbstractResolverBaselineCorrector? looks the same to me now

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/AbstractResolverBaselineCorrector.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import io.github.mzmine.util.MemoryMapStorage;
+import java.util.Arrays;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Baseline correction that uses feature resolver as intermediate step
+ */
+public abstract class AbstractResolverBaselineCorrector extends AbstractBaselineCorrector {
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/AbstractResolverBaselineCorrector.java:42]: The chang corrector algorithms odes not use / implement the feature resolver step - can also be changed. 
I was not sure if all baseline correctors will use the resolver.

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/BaselineDataBuffer.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+import io.github.mzmine.util.collections.IndexRange;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Logger;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+
+public class BaselineDataBuffer {
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/BaselineDataBuffer.java:41]: the logic of creating the buffers and extending the size were scattered in multiple classes and the handling was even different in a class

[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/BaselineDataBuffer.java:41]: Now it is only one variable in a class and not 4

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/BaselineDataBuffer.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+      yAlreadySet = true;
+    }
+
+    ensureCapacity(numValues);
+    extractRtValues(timeSeries);
+    clearRemovedPeaksBuffers();
+
+    indicesOfInterest.clear();
+    indicesOfInterest.add(0);
+    indicesOfInterest.add(numValues - 1);
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/BaselineDataBuffer.java:85]: good question I will check

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/UnivariateBaselineCorrector.java (hunk 2 above)
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/UnivariateBaselineCorrector.java:94]: good catch. This was the idea to pull this up but somehow forgot to delete the override

Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/UnivariateBaselineCorrector.java:94]: I was also able to remove the PolynomialBaselineCorrection.subSampleAndCorrect which was very similar to UnivariateBaselineCorrection
Found out that PolynomialFunction was using apache.math and not math3 now they share the same UnivariateFuntion interface and I could just implement the UnivariateBaselineCorrector

File: mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/AbstractBaselineCorrector.java [hunk 6]
```diff
@@ -163,4 +125,20 @@ public void setPreview(boolean preview) {
...
+    return buffer.yBuffer();
+  }
+
+  public double[] xBufferRemovedPeaks() {
+    return buffer.xBufferRemovedPeaks();
+  }
+
+  public double[] yBufferRemovedPeaks() {
+    return buffer.yBufferRemovedPeaks();
+  }
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/AbstractBaselineCorrector.java:143]: was initially easier to change by just adding the () to each use. 
But I agree makes more sense to pass the buffer instead.

[the remark this answers is not in the document]
Comment (robinschmid) [reply mzmine-community/src/main/java/io/github/mzmine/modules/dataprocessing/featdet_baselinecorrection/AbstractBaselineCorrector.java:143]: all removed. Made it much easier and nicer now

**Comment by robinschmid** [comment]
ready from my side

**Pull request merged**